netam/whichmut_dataset.py

```diff
@@ -0,0 +1,673 @@
+modeling.
+
+## Sparse Data Structure Format
+
+The sparse neutral rates format is designed for memory efficiency when dealing with
+large-scale sequence datasets. Instead of storing full 65x65 transition matrices
+(most entries are zero), it stores only the non-zero single-nucleotide mutation rates.
+
+### Core Data Structure
+
+The sparse format uses a dictionary with three tensors:
+
+```python
+sparse_neutral_rates = {
+    'indices': torch.LongTensor,     # Shape: (N, L, max_mutations, 2)
+    'values': torch.Tensor,          # Shape: (N, L, max_mutations)
+    'n_possible_mutations': torch.LongTensor  # Shape: (N, L)
+}
+```
+
+### Components Explained
+
+1. **`indices` tensor (N, L, max_mutations, 2)**:
+   - Stores [parent_codon_idx, child_codon_idx] pairs for each possible mutation
+   - Uses full codon indices (0-64 range), not compressed indices
+   - `max_mutations` is always 9 (3 positions × 3 alternative bases per position)
+   - Only the first `n_possible_mutations[seq, pos]` entries are valid per position
+
+2. **`values` tensor (N, L, max_mutations)**:
+   - Contains the actual neutral rates λ_{j,c->c'} for each mutation
+   - Aligned with the `indices` tensor - same indexing scheme
+   - Only the first `n_possible_mutations[seq, pos]` entries are valid per position
+
+3. **`n_possible_mutations` tensor (N, L)**:
+   - Indicates how many valid single-nucleotide mutations are stored per position
+   - Typically 7-9 mutations per codon (after filtering stop codons)
+   - The name emphasizes these are *possible* mutations, not observed mutations
+
+### FUNCTIONAL_CODON_SINGLE_MUTATIONS Lookup Table
+
+The sparse data structure is built using `FUNCTIONAL_CODON_SINGLE_MUTATIONS` from
+`netam.codon_table`, which has the structure:
+
+```python
+FUNCTIONAL_CODON_SINGLE_MUTATIONS = {
+    parent_codon_idx: [
+        (child_codon_idx, nt_position, new_base),
+        ...
+    ]
+}
+```
+
+**Key properties**:
+- Excludes mutations to stop codons (only functional codons included)
+- Each parent codon maps to 7-9 possible single-nucleotide mutations
+- `nt_position` is 0, 1, or 2 (position within the codon)
+- `new_base` is the replacement nucleotide
+
+### Example Sparse Data
+
+For a codon AAA (index 0) at one sequence position:
+
+```python
+# AAA has 8 functional single mutations (excluding stop codons)
+indices[0, 0, :8, :] = [
+    [0, 1],   # AAA -> AAC (Lys -> Asn)
+    [0, 2],   # AAA -> AAG (Lys -> Lys, synonymous)
+    [0, 4],   # AAA -> ACA (Lys -> Thr)
+    [0, 8],   # AAA -> AGA (Lys -> Arg)
+    [0, 16],  # AAA -> CAA (Lys -> Gln)
+    [0, 32],  # AAA -> GAA (Lys -> Glu)
+    [0, 48],  # AAA -> TAA would be stop - excluded
+    [0, 56],  # AAA -> TCA (Lys -> Ser)
+    # ... (8th mutation)
+]
+
+values[0, 0, :8] = [0.01, 0.015, 0.012, ...]  # Corresponding neutral rates
+n_possible_mutations[0, 0] = 8  # 8 valid mutations stored
+```
+
+### Memory Efficiency
```

Comment:
I'd suggest removing this section (just the facts). Also I don't think that the relevant comparison is to N × L × 65², but rather N × L × 65 . This appears above.

Reply:
(It's inconsistent with the true dense format described below)